test/over_react/util/safe_render_manager/safe_render_manager_test.dart

```diff
@@ -0,0 +1,707 @@
+          renderManager.render(Wrapper()());
+
+          expect(renderManager.contentRef, isNotNull);
+          expect(renderManager.contentRef, const isInstanceOf<WrapperComponent>());
+        });
+
+        test('by chaining any existing callback ref', () {
+          WrapperComponent existingWrapperRef;
+
+          renderManager.render((Wrapper()..ref = ((ref) => existingWrapperRef = ref))());
+
+          expect(renderManager.contentRef, isNotNull);
+          expect(existingWrapperRef, same(renderManager.contentRef));
+        });
+      });
+    });
+
+    group('tryUnmount()', () {
+      group('unmounts the rendered component', () {
+        setUp(() {
+          renderManager.render(Wrapper()());
+          expect(mountNode.children, isNotEmpty);
+        });
+
+        test('', () async {
+          renderManager.tryUnmount();
+          expect(mountNode.children, isEmpty);
+        });
+
+        test('and calls the provided callback when complete', () async {
+          renderManager.render(Wrapper()());
+
+          expect(mountNode.children, isNotEmpty);
+
+          final onUnmountCompleter = new Completer();
+
+          renderManager.tryUnmount(onMaybeUnmounted: bind1Guarded((isUnmounted) {
+            expect(mountNode.children, isEmpty);
+            expect(isUnmounted, isTrue);
+
+            onUnmountCompleter.complete();
+          }));
+
+          await onUnmountCompleter.future;
+        });
+      });
+
+      test('invokes the provided callback immediately when nothing has been rendered', () async {
+        expect(mountNode.children, isEmpty, reason: 'test setup sanity check');
+
+        bool onUnmountCalledSynchronously = false;
+
+        renderManager.tryUnmount(onMaybeUnmounted: bind1Guarded((isUnmounted) {
+          expect(isUnmounted, isTrue);
+          onUnmountCalledSynchronously = true;
+        }));
+
+        expect(onUnmountCalledSynchronously, isTrue);
+      });
+    });
+
+    group('automatically attaches and detached the mount node', () {
+      setUp(() async {
+        // Clean up the manager from the above setUp block.
+        await renderManager?.dispose();
+        renderManager = null;
+      });
+
+      test('when autoAttachMountNode is true', () {
+        renderManager = new SafeRenderManager(autoAttachMountNode: true);
+        expect(document.contains(renderManager.mountNode), isFalse, reason: 'test setup check');
+
+        renderManager.render(Dom.div()());
+        expect(renderManager.mountNode.parent, document.body);
+
+        renderManager.tryUnmount();
+        expect(renderManager.mountNode.parent, isNull);
+      });
+
+      test('unless autoAttachMountNode is false', () {
+        renderManager = new SafeRenderManager(autoAttachMountNode: false);
+        expect(renderManager.mountNode.parent, isNull, reason: 'test setup check');
+
+        renderManager.render(Dom.div()());
+        expect(renderManager.mountNode.parent, isNull);
```

Comment:
If `mountNode` is not specified, and `autoAttachMountNode` is false... isn't that an invalid configuration? What would it render into?

I realize this test is not really testing that - I'm just saying that the conditions of this test made me think this probably shouldn't be supported.

Reply:
No, it's a valid configuration. The manager defaults to a new, empty div if one isn't specified, and the consumer would be responsible for attaching/detaching it from the DOM.